feat: Adds support for region & customer_endpoint_dns_name in privatelink (#1982)
mongodb/terraform-provider-mongodbatlas: website/docs/r/privatelink_endpoint_service_data_federation_online_archive.html.markdown

>   endpoint_id = "<PRIVATE-ENDPOINT-SERVICE-ID>"

Reviewer: ```suggestion
  endpoint_id = "vpce-3bf78b0ddee411ba1"
```

Author: Changes made in 7c55d85 ✅

  provider_name = "AWS"
  comment = "Test"
  region        = "US_EAST_1"

>   customer_endpoint_dns_name = aws_vpc_endpoint.<AWS-RESOURCE_NAME>.dns_entry[0].dns_name

Reviewer: ```suggestion
  customer_endpoint_dns_name = "vpce-046cf43c79424d4c9-nmls2y9k.vpce-svc-0824460b72e1a420e.us-east-1.vpce.amazonaws.com"
```

Author: Changes made in 7c55d85 ✅

}
```

## Argument Reference

* `project_id` (Required) - Unique 24-hexadecimal digit string that identifies your project. 
* `endpoint_id` (Required) - Unique 22-character alphanumeric string that identifies the private endpoint. See [Atlas Data Lake supports Amazon Web Services private endpoints using the AWS PrivateLink feature](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint:~:text=Atlas%20Data%20Lake%20supports%20Amazon%20Web%20Services%20private%20endpoints%20using%20the%20AWS%20PrivateLink%20feature).
* `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
* `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about timeouts [here](https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts).

> * `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).

Reviewer: ```suggestion
* `region` - (Optional) Human-readable label to identify the region of VPC endpoint.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).
```

> * `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`

Reviewer: ```suggestion
* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`
```

> * `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`

Reviewer: ```suggestion
* `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name.
```

Author: Changes made in 7c55d85 ✅

* `endpoint_id` (Required) - Unique 22-character alphanumeric string that identifies the private endpoint. See [Atlas Data Lake supports Amazon Web Services private endpoints using the AWS PrivateLink feature](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint:~:text=Atlas%20Data%20Lake%20supports%20Amazon%20Web%20Services%20private%20endpoints%20using%20the%20AWS%20PrivateLink%20feature).
* `provider_name` (Required) - Human-readable label that identifies the cloud service provider. 
* `timeouts`- (Optional) The duration of time to wait for Private Endpoint Service to be created or deleted. The timeout value is definded by a signed sequence of decimal numbers with an time unit suffix such as: `1h45m`, `300s`, `10m`, .... The valid time units are:  `ns`, `us` (or `µs`), `ms`, `s`, `m`, `h`. The default timeout for Private Endpoint create & delete is `2h`. Learn more about timeouts [here](https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts).

> * `region` - (Optional) Physical location of your MongoDB cluster. The region you choose can affect network latency for clients accessing your databases.  Requires the **Atlas region name**, see the reference list for [AWS](https://docs.atlas.mongodb.com/reference/amazon-aws/), [GCP](https://docs.atlas.mongodb.com/reference/google-gcp/), [Azure](https://docs.atlas.mongodb.com/reference/microsoft-azure/).

Reviewer 2: there would be two similar doc under `website/docs/d` for data sources that need to be updated as well

Author: Added. Wasn't sure whether or not to include (Optional) in the data sources

Reviewer 2: depends, is this attribute required to be provided by the user when calling this data source? what happens if the user doesn't provide it? how is the API call made then?
Did you try using the data source to get a better idea around usage?

Author: It is never used when reading.
Maybe better to mark is as `Computed`?

Reviewer 2: yes should be computed in data source, I added a comment on the data source file itself so there's no confusion, let's resolve this one

> * `customer_endpoint_dns_name` - (Optional) Human-readable label to identify customer's VPC endpoint DNS name. Must follow regex: `vpce-[a-z0-9]{17}-[a-z0-9]{8}.vpce-svc-[a-z0-9]{17}.(.*).vpce.amazonaws.com`

Reviewer 3: is this regex imposed by us or from the backend? if so, could we reference the official documentation where this is validated as well?

Author: Yes, I learned this the "hard" way 😅 
I had to dig into the code to find it. The [API docs](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Data-Federation/operation/createDataFederationPrivateEndpoint)  didn't say anything.
I messaged Zuhair about it, but I didn't get a reply yet

Reviewer: Made a few suggestions directly in this PR. Cleaned up example to make sure reader is aware of appriorapate formats for `customer_endpoint_dns_name` and `endpoint_id`. Also updated docs definitions to be more inline with Admin API.

Author: Changes made in 7c55d85 ✅

> * `comment` - (Optional) Human-readable string to associate with this private endpoint.

Reviewer 2: did we confirm if it's okay to allow users to update comment?

Reviewer 4: if we support the update for the comment, might be worth add a note at the start of the docs mentioning that the update operation is limited to the comment attribute.

Author: I added a note now. You can resolve if it looks ok

Author: Changes made in 7c55d85 ✅
